### PR TITLE
Allow custom cookie consent value

### LINKED ISSUE #40164: Allow customizing the cookie consent value
## Background and Motivation

Currently, the cookie consent validation performs an equality comparison against the constant cookie value `yes`. It will add more flexibility if the value can be customized, especially when supporting an already existing cookie consent.

I already made the changes in my fork and added some tests, you can see them right here 68180e5d85730ddcc77d06176a1af300fe668afa.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Builder;

public class CookiePolicyOptions
{
    /// <summary>
    /// Gets or sets the <see cref="CookieBuilder"/> that is used to track if the user consented to the
    /// cookie use policy.
    /// </summary>
    public CookieBuilder ConsentCookie { get; set; } = new CookieBuilder()
    {
        Name = ".AspNet.Consent",
        Expiration = TimeSpan.FromDays(365),
        IsEssential = true,
    };

+   /// <summary>
+   /// Gets or sets the consent value that is used to track if the user consented to the
+   /// cookie use policy. The default is yes.
+   /// </summary>
+   public string ConsentCookieValue { get; set; } = "yes";
}
```

```diff
namespace Microsoft.AspNetCore.CookiePolicy;

internal class ResponseCookiesWrapper : IResponseCookies, ITrackingConsentFeature
{
-   private const string ConsentValue = "yes";
+   private const string DefaultConsentValue = "yes";
    private readonly ILogger _logger;
    private bool? _isConsentNeeded;
    private bool? _hasConsent;

    public ResponseCookiesWrapper(HttpContext context, CookiePolicyOptions options, IResponseCookiesFeature feature, ILogger logger)
    {
        Context = context;
        Feature = feature;
        Options = options;
        _logger = logger;

+       if (string.IsNullOrWhiteSpace(Options.ConsentCookieValue))
+       {
+           Options.ConsentCookieValue = DefaultConsentValue;
+       }
    }

    public bool HasConsent
    {
        get
        {
            if (!_hasConsent.HasValue)
            {
                var cookie = Context.Request.Cookies[Options.ConsentCookie.Name!];
-               _hasConsent = string.Equals(cookie, ConsentValue, StringComparison.Ordinal);
+               _hasConsent = string.Equals(cookie, Options.ConsentCookieValue, StringComparison.Ordinal);
                _logger.HasConsent(_hasConsent.Value);
            }

            return _hasConsent.Value;
        }
    }

    public void GrantConsent()
    {
        if (!HasConsent && !Context.Response.HasStarted)
        {
            var cookieOptions = Options.ConsentCookie.Build(Context);
            // Note policy will be applied. We don't want to bypass policy because we want HttpOnly, Secure, etc. to apply.
-           Append(Options.ConsentCookie.Name!, ConsentValue, cookieOptions);
+           Append(Options.ConsentCookie.Name!, Options.ConsentCookieValue, cookieOptions);
            _logger.ConsentGranted();
        }
        _hasConsent = true;
    }

    public string CreateConsentCookie()
    {
        var key = Options.ConsentCookie.Name;
-       var value = ConsentValue;
+       var value = Options.ConsentCookieValue;
        var options = Options.ConsentCookie.Build(Context);

        Debug.Assert(key != null);
        ApplyAppendPolicy(ref key, ref value, options);

        var setCookieHeaderValue = new Net.Http.Headers.SetCookieHeaderValue(
            Uri.EscapeDataString(key),
            Uri.EscapeDataString(value))
        {
            Domain = options.Domain,
            Path = options.Path,
            Expires = options.Expires,
            MaxAge = options.MaxAge,
            Secure = options.Secure,
            SameSite = (Net.Http.Headers.SameSiteMode)options.SameSite,
            HttpOnly = options.HttpOnly
        };

        return setCookieHeaderValue.ToString();
    }
}
```

## Usage Examples

``` C#
services.Configure<CookiePolicyOptions>(options =>
{
    options.CheckConsentNeeded = context => true;
    options.ConsentCookieValue = "true";
});
```

## Alternative Designs

None at the moment.

## Risks

I don't see any breaking changes since the default value for the cookie consent will remain as `yes`.

### PR DESCRIPTION
**Pull request by daviddesmet**
# Allow customizing the cookie consent value

Allows setting a custom value for the cookie consent.

## Description

Currently, the cookie consent validation performs an equality comparison against the constant cookie value `yes`. It will add more flexibility if the value can be customized, especially when supporting an already existing cookie consent.

Fixes #40164

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=40213) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=40213) before we can accept your contribution.<br/><br/>:x: daviddesmet [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=40213)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=40213) it.</sub>

File: src/Security/CookiePolicy/src/ResponseCookiesWrapper.cs [hunk 1]
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.AspNetCore.CookiePolicy;
 
 internal class ResponseCookiesWrapper : IResponseCookies, ITrackingConsentFeature
 {
-    private const string ConsentValue = "yes";
+    private const string DefaultConsentValue = "yes";
```
Comment (pranavkm) [inline src/Security/CookiePolicy/src/ResponseCookiesWrapper.cs:15]:
```suggestion
```

**Comment by daviddesmet** [comment]
@pranavkm The test `CreateConsentCookieMatchesGrantConsentCookieWhenCookieValueIsMissing()` will need to be updated since the `ConsentCookieValue` setter no longer accepts an empty value.

**Comment by pranavkm** [comment]
Thanks. If you feel like it, it would be nice to add a test for CookieOptions to check it does not allow null / empty values for the consent value.

**Review by Tratcher — APPROVED** [review]

**Comment by daviddesmet** [comment]
@pranavkm Sure, I will add it!

**Comment by daviddesmet** [comment]
@pranavkm @Tratcher Unit Test added when consent value is missing.

**Comment by daviddesmet** [comment]
@pranavkm The tests that failed didn't like 'ExceptionAssert.ThrowsArgument(Action, string, string)' :(

**Pull request merged**